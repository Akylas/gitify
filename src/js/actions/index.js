import { apiRequest, apiRequestAuth } from '../utils/api-requests';

import Constants from '../utils/constants';

export function makeAsyncActionSet(actionName) {
  return {
    REQUEST: actionName + '_REQUEST',
    SUCCESS: actionName + '_SUCCESS',
    FAILURE: actionName + '_FAILURE'
  };
}

// Authentication

export const LOGIN = makeAsyncActionSet('LOGIN');
export function loginUser(code) {
  return (dispatch, getState) => {

    const url = 'https://github.com/login/oauth/access_token';
    const method = 'POST';
    const data = {
      'client_id': Constants.CLIENT_ID,
      'client_secret': Constants.CLIENT_SECRET,
      'code': code
    };

    dispatch({type: LOGIN.REQUEST});

    return apiRequest(url, method, data)
      .then(function (response) {
        dispatch({type: LOGIN.SUCCESS, payload: response.data});
      })
      .catch(function (error) {
        dispatch({type: LOGIN.FAILURE, payload: error.response.data});
      });

  };
};

export const LOGOUT = 'LOGOUT';
export function logout() {
  return { type: LOGOUT };
};


// Notifications

export const NOTIFICATIONS = makeAsyncActionSet('NOTIFICATIONS');
export function fetchNotifications() {
  return (dispatch, getState) => {

    const participating = getState().settings.get('participating');
    const url = `https://api.github.com/notifications?participating=${participating}`;
    const method = 'GET';
    const token = getState().auth.get('token');

    dispatch({type: NOTIFICATIONS.REQUEST});

    return apiRequestAuth(url, method, token)
      .then(function (response) {
        dispatch({type: NOTIFICATIONS.SUCCESS, payload: response.data});
      })
      .catch(function (error) {
        dispatch({type: NOTIFICATIONS.FAILURE, payload: error.response.data});
      });

  };
};


// Single Notification

export const MARK_NOTIFICATION = makeAsyncActionSet('MARK_NOTIFICATION');
export function markNotification(id) {
  return (dispatch, getState) => {

    const url = `https://api.github.com/notifications/threads/${id}`;
    const method = 'PATCH';
    const token = getState().auth.get('token');

    dispatch({type: MARK_NOTIFICATION.REQUEST});

    return apiRequestAuth(url, method, token, {})
      .then(function (response) {
        dispatch({type: MARK_NOTIFICATION.SUCCESS, payload: response.data, meta: { id }});
      })
      .catch(function (error) {
        dispatch({type: MARK_NOTIFICATION.FAILURE, payload: error.response.data});
      });
  };
};


// Repo's Notification

export const MARK_REPO_NOTIFICATION = makeAsyncActionSet('MARK_REPO_NOTIFICATION');
export function markRepoNotifications(loginId, repoId, repoFullName) {
  return (dispatch, getState) => {

    const url = `https://api.github.com/repos/${loginId}/${repoId}/notifications`;
    const method = 'PUT';
    const token = getState().auth.get('token');

    dispatch({type: MARK_REPO_NOTIFICATION.REQUEST});

    return apiRequestAuth(url, method, token, {})
      .then(function (response) {
        dispatch({type: MARK_REPO_NOTIFICATION.SUCCESS, payload: response.data, meta: { repoFullName, repoId }});
      })
      .catch(function (error) {
        dispatch({type: MARK_REPO_NOTIFICATION.FAILURE, payload: error.response.data});
      });
  };
};


// Starred

export const HAS_STARRED = makeAsyncActionSet('HAS_STARRED');
export function checkHasStarred() {
  return (dispatch, getState) => {

    const url = `https://api.github.com/user/starred/${Constants.REPO_SLUG}`;
    const method = 'GET';
    const token = getState().auth.get('token');

    dispatch({type: HAS_STARRED.REQUEST});

    return apiRequestAuth(url, method, token)
      .then(function (response) {
        dispatch({type: HAS_STARRED.SUCCESS, payload: response.data});
      })
      .catch(function (error) {
        dispatch({type: HAS_STARRED.FAILURE, payload: error.response.data});
      });

  };
};


// Search

export const SEARCH_NOTIFICATIONS = 'SEARCH_NOTIFICATIONS';
export const CLEAR_SEARCH = 'CLEAR_SEARCH';
export function searchNotifications(query) {
  return {
    type: SEARCH_NOTIFICATIONS,
    query: query
  };
};

export function clearSearch() {
  return {
    type: CLEAR_SEARCH
  };
};


// Settings

export const UPDATE_SETTING = 'UPDATE_SETTING';
export function updateSetting(setting, value) {
  return {
    type: UPDATE_SETTING,
    setting: setting,
    value: value
  };
};
