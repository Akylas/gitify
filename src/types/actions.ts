export const TOGGLE_SETTINGS_MODAL = 'TOGGLE_SETTINGS_MODAL';

export interface ToggleSettingsModalAction {
  type: typeof TOGGLE_SETTINGS_MODAL;
}
