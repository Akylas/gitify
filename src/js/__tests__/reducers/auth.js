import { Map, fromJS } from 'immutable';
import { expect } from 'chai';

import reducer from '../../reducers/auth';
import { LOGIN, LOGOUT } from '../../actions';

describe('reducers/auth.js', () => {
  const initialState = fromJS({
    response: {},
    token: null,
    isFetching: false,
    failed: false
  });

  it('should return the initial state', () => {

    expect(reducer(undefined, {})).to.eql(initialState);

  });

  it('should handle LOGIN.REQUEST', () => {

    const action = {
      type: LOGIN.REQUEST
    };

    expect(reducer(undefined, action)).to.eql(
      initialState
        .set('isFetching', true)
        .set('failed', false)
        .set('response', Map())
        .set('token', null)
    );

  });

  it('should handle LOGIN.SUCCESS', () => {

    const fakeState = fromJS({
      isFetching: true,
      failed: false,
      response: {},
      token: null
    });

    expect(reducer(fakeState, {}).get('token')).to.be.null;

    const action = {
      type: LOGIN.SUCCESS,
      payload: {
        access_token: '123HELLOWORLDTOKEN'
      }
    };

    expect(reducer(fakeState, action)).to.eql(
      initialState
      .set('isFetching', false)
      .set('token', action.payload.access_token)
    );

    expect(reducer(fakeState, action).token).to.not.be.null;

  });

  it('should handle LOGIN.FAILURE', () => {

    const fakeState = fromJS({
      isFetching: true,
      failed: false,
      response: {},
      token: null
    });

    expect(reducer(fakeState, {}).get('token')).to.be.null;

    const action = {
      type: LOGIN.FAILURE,
      payload: {msg: 'Failed to login.'}
    };

    expect(reducer(fakeState, action)).to.eql(
      initialState
        .set('isFetching', false)
        .set('failed', true)
        .set('response', Map(action.payload))
    );

    expect(reducer(fakeState, action).get('token')).to.be.null;

  });

  it('should handle LOGOUT', () => {

    const fakeState = fromJS({
      isFetching: false,
      failed: false,
      response: {},
      token: 'LOGGEDINTOKEN'
    });

    expect(reducer(fakeState, {}).get('token')).to.not.be.null;

    const action = {
      type: LOGOUT
    };

    expect(reducer(fakeState, action)).to.eql(
      initialState
        .set('token', null)
        .set('response', null)
    );

    expect(reducer(fakeState, action).get('token')).to.be.null;

  });

});
