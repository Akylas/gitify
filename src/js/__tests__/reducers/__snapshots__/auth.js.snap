// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/auth.js should handle LOGIN.FAILURE 1`] = `
Object {
  "failed": true,
  "isFetching": false,
  "response": Object {
    "msg": "Failed to login.",
  },
  "token": null,
}
`;

exports[`reducers/auth.js should handle LOGIN.REQUEST 1`] = `
Object {
  "enterpriseAccounts": Array [],
  "failed": false,
  "isFetching": true,
  "response": Object {},
  "token": null,
}
`;

exports[`reducers/auth.js should handle LOGIN.SUCCESS - enterprise 1`] = `
Object {
  "enterpriseAccounts": Array [
    Object {
      "hostname": "github.gitify.io",
      "token": "123HELLOWORLDTOKEN",
    },
  ],
  "failed": false,
  "isFetching": false,
  "response": Object {},
  "token": null,
}
`;

exports[`reducers/auth.js should handle LOGIN.SUCCESS - github.com 1`] = `
Object {
  "enterpriseAccounts": Array [],
  "failed": false,
  "isFetching": false,
  "response": Object {},
  "token": "123HELLOWORLDTOKEN",
}
`;

exports[`reducers/auth.js should handle LOGOUT 1`] = `
Object {
  "enterpriseAccounts": Array [],
  "failed": false,
  "isFetching": false,
  "response": null,
  "token": null,
}
`;

exports[`reducers/auth.js should return the initial state 1`] = `
Object {
  "enterpriseAccounts": Array [],
  "failed": false,
  "isFetching": false,
  "response": Object {},
  "token": null,
}
`;
